fix race condition, output non-inverted, filegrp directory, pageId

bertsky opened the pull request:
Mind you: the tests currently _do not_ pass, but only because we now expect the positive (not the inverted) images as results – see OCR-D/assets#36

Commit:
fix race condition in processor: between save_mets...

- in 1st cmd of pipeline when doing `ocrd workspace find --download`
- in 2nd cmd of pipeline when doing `ocrd workspace add`

...by not using a pipeline at all!

Commit:
add JPG MIME type, show proper warning if MIME type is unknown

Commit:
invert the binarized images (so they come out positive)

Commit:
params: add 2 missing binarization methods

Commit:
produce files under output file group's directory:

- abandon filename extension guesswork based on MIME type
  (just name the file by its ID plus new extension)
- re-use the old file ID, but replace input with output file group
- write the file into the output file group directory
  (and create it, if it did not exist)

Commit:
pass on pageId, if it exists

Commit:
improve makefile

Commit:
tests: adjust to new assets, adjust to new behaviour

bertsky commented:
@wrznr @kba please review!

wrznr commented on ocrd-olena-binarize:
```diff
@@ -6,52 +6,60 @@ which ocrd >/dev/null 2>/dev/null || { echo "ocrd not in \$PATH. Panicking"; exi
+        --download))
+    for csv in "${files[@]}"; do
         # Parse comma separated fields
         local IFS=$'\t'
```
What is `ISF`? It is defined twice but never accessed.

bertsky replied:
`IFS` is bash's _internal field separator_ variable. It is needed here to mark the boundaries of array elements (newline vs tab). First, we only want the newlines when we init the `files` array. Then inside one line we only want tabs to init the `fields` array.

wrznr commented on ocrd-olena-binarize:
```diff
@@ -6,52 +6,60 @@ which ocrd >/dev/null 2>/dev/null || { echo "ocrd not in \$PATH. Panicking"; exi
+            -g $in_pageId    \
+            -m image/png     \
+            -i "$out_id"     \
             "$out_fpath"
```
No sleep? 😃

bertsky replied:
No sleep not necessary no longer! (Didn't help anyway, it was only a matter of how often clashes occured.)

wrznr commented on ocrd-tool.json:
```diff
@@ -4,19 +4,25 @@
     "ocrd-olena-binarize": {
       "executable": "ocrd-olena-binarize",
-      "description": "OLENA's binarization algos for OCR-D",
+      "description": "OLENA's binarization algos for OCR-D (on page-level)",
```
What would have to be done to expose them on non-page levels?

bertsky replied:
Much more I am afraid. We would have to parse PAGE inputs down to the requested hierarchy level, download the `AlternativeImage/@filename` from the workspace, process that, and reference the new file in another `AlternativeImage`, and finally serialise the PAGE again – all in bash.

kba replied:
We could implement the cropping to specific regions in python and expose it as a CLI. It's still a lot of work to design, implement and test though.

wrznr commented:
Fixes #1 !

kba commented on Makefile:
```diff
@@ -38,7 +39,7 @@ olena-git:
 
 deps-ubuntu:
-	sudo apt install libmagick++-dev `grep -q 18.04 /etc/*release || libtesseract3-dev`
+	sudo apt install libmagick++-dev libboost-dev `grep -q 18.04 /etc/*release || echo libtesseract-dev`
```
clever 🥇

kba commented on ocrd-olena-binarize:
```diff
@@ -6,52 +6,60 @@ which ocrd >/dev/null 2>/dev/null || { echo "ocrd not in \$PATH. Panicking"; exi
 
         # Convert PBM to PNG
-        convert  "$out_fpath.pbm" "$out_fpath"
+        convert "$out_fpath.pbm" -negate "$out_fpath"
```
IIRC not all the binarization algos output positive images. Probably best to have a minimalist test for every algo.

bertsky replied:
Oh, really? Then it's probably better to determine which do and which don't just once (and by looking at the code), and either make the negation conditional on the name of the method, or use the specialized binaries (instead of `scribo-cli`) altogether.

bertsky replied:
All binarizations I tried (more than those currently exposed in ocrd-tool.json) came out negative. I did not find a common code responsible for this though. Rather, things like [this]... I think we can called that "resolved enough", right?

kba reviewed (approved):
Great work! I'll update the makefile so the makefile-parser help mechanism works and merge it. Thank you!

bertsky commented:
Ah, wait please. I am in the middle of another fix for the makefile.

bertsky commented:
Nearly done. (We need to patch Olena's configure script so it recognizes the boost installation correctly. Is it ok if I include a patch in the repo, which I apply in the makefile right after unpacking the tarball?)

kba commented:
> Is it ok if I include a patch in the repo, which I apply in the makefile right after unpacking the tarball?)

:+1:

Commit:
fix olena build:

- patch Olena's configure to properly detect libboost
  (cpp ran with line comments enabled, which prevented parsing the results)
- add missing dependency for Ubuntu: GraphicsMagick++
  (unfortunately, configure does not regard this as critical)
- in checking whether olena is built/installed already, do not just test
  scribo-cli, but also one of its tools (sauvola)
- when building olena, depend on success of configure, and
  use proper recursive make
- make deps-ubuntu independent (only for Docker or user)
- make install dependent on deps
- make test dependent on install

bertsky commented:
@kba done! I hope this does not make your makefile-parsing more difficult.

wrznr commented:
Travis complains about missing `neato` which is part of the graphviz package. Could you add it to the dependencies?

Commit:
makefile: need graphviz too (Travis)

bertsky commented:
@wrznr Done!

bertsky commented:
Can someone please fix OCR-D/assets#36?

Commit:
travis: fix deps (hopefully)

wrznr commented:
`sudo` for `apt`?

bertsky commented:
Right, sorry! (I was thinking dockerfile...)

bertsky commented:
thumbs crossed!

wrznr commented:
It is `libboost-all-dev`!

wrznr commented:
It is `libgraphicsmagick++-dev`!

kba commented:
Switch to CircleCI/Docker based on a less ancient distro?

wrznr commented:
But `libtesseract-dev` should work :cry:

bertsky commented:
I don't understand. I do have a 14.04 for reference here, and there both packages are named as currently in `.travis.yml`...

bertsky commented:
Maybe an `apt update` is missing?

bertsky commented:
Indeed!

bertsky commented:
Oh no. This looks pretty much like OCR-D/core#96 ...

bertsky commented:
I think we could replace the `ocrd workspace clone -a "$assets/scribo-test/data/mets.xml" "$workspace_dir"` in test.sh with a stupid `cp -r` – but I really prefer fixing this in core!

Commit:
test: avoid core#96

bertsky commented:
At last!

Commit:
exposed 3 more binarization methods

wrznr commented:
Great. Merge it!

kba commented on ocrd-tool.json:
```diff
@@ -22,7 +22,7 @@
           "type": "string",
           "required": true,
-          "enum": ["sauvola", "sauvola-ms", "sauvola-ms-fg", "sauvola-ms-split", "kim", "wolf"]
+          "enum": ["sauvola", "sauvola-ms", "sauvola-ms-fg", "sauvola-ms-split", "kim", "wolf", "niblack", "singh", "otsu"]
```
I could extend the scribo-test assets with samples for kim, wolf, niblack, singh, otsu. As they are, these are just smoke tests, comparing file sizes but for consistency's sake... Should I?

bertsky replied:
No opinion here. But maybe we could include the self-tests of scribo into our Travis configuration.
